### PR TITLE
feat(calculator): add sequential bayes factor stopping rule for always-valid inference

### PR DESCRIPTION
**Pull request by cre8ivejp**
## Summary

- Adds a sequential Bayes Factor stopping rule to the experiment calculator
  so that peeking at the dashboard and stopping early does not inflate the
  false-positive rate.
- Two new proto fields on `VariationResult` expose the per-variation BF for
  CVR (`cvr_sequential_bayes_factor`) and value/user
  (`value_sequential_bayes_factor`). Two new booleans on
  `GoalResult.Summary` (`cvr_safe_to_stop`, `value_safe_to_stop`) encode
  the stopping decision.
- The existing `ProbBeatBaseline` column and `BestVariations` list are
  unchanged; the BF gate is additive. A subsequent UI PR will reframe the
  confidence banner to surface the dual signal.

## Motivation

The current verdict rule (`ProbBeatBaseline > 0.95`) is calibrated for a
single look. A user who checks daily and stops at first crossing is running
an implicit sequential test with a single-look threshold — the true
false-positive rate over 28 daily looks can reach 25–35% for a null
experiment. This PR closes that gap for both CVR and value metrics.

### CVR — Beta-Binomial Bayes Factor

H₁: θ_A ~ Beta(1,1), θ_B ~ Beta(1,1) independently
H₀: θ_A = θ_B = θ,  θ ~ Beta(1,1)

BF₁₀ = B(s_A+1, n_A−s_A+1) · B(s_B+1, n_B−s_B+1) / B(s_A+s_B+1, n_A+n_B−s_A−s_B+1)

Closed-form, computed in log-space via `math.Lgamma`. No Stan call required.
The BF sequence is a martingale under H₀:

> P(max_t BF(t) ≥ 20 | H₀, Bayesian-averaged) ≤ 1/20 = 5%

### Value/user — Normal-Normal BF on the standardized difference

A plain NIG marginal-likelihood BF with μ₀=0 suffers from Bartlett's paradox:
when both arm means are far from 0, H₀ is systematically preferred because it
pays the prior-mean penalty once instead of twice. The t-statistic approach
tests the *difference* δ = μ_B − μ_A with a fixed unit-information prior,
avoiding this problem:

BF₁₀ = sqrt(1/(1 + n_eff·r²)) · exp(t² · n_eff·r² / (2·(1 + n_eff·r²)))

where:
- t = (mean_B − mean_A) / sqrt(var_A/n_A + var_B/n_B)  (Welch t-statistic)
- n_eff = n_A·n_B / (n_A+n_B)
- r = 1.0  (fixed constant — must not be data-derived)

E[BF | H₀, t~N(0,1)] = 1 exactly in the large-n limit. For finite n the
t-distribution tails make it mildly anti-conservative (~1–2% FPR vs ~0.5%
for CVR); validated at n_per_arm ≥ 30.

### Key caveats

- **Bayesian-averaged, not per-θ worst-case.** The multi-θ simulation audits
  θ ∈ {0.01, 0.05, 0.10, 0.30, 0.50}; per-θ warn level is 12%.
- **Per-comparison, not family-wise.** With k treatment arms the family-wise
  FPR inflates ~linearly. Full multiple-comparisons handling is deferred.
- **Power cost.** The sequential rule costs ~15–25% more average n for small
  effects vs a fixed-horizon test; for large effects it terminates earlier.
  `safe_to_stop = false` means "more data needed", not "inconclusive."

## Changes

| File | What |
|---|---|
| `sequential_bayes_factor.go` | New: `cvrBayesFactor`, `valueBayesFactor`, `fillSequentialBayesFactors`, `computeSafeToStop`, threshold constant |
| `sequential_bayes_factor_test.go` | New: 18 unit tests + 2 FPR simulations (CVR multi-θ, value multi-scenario) |
| `variation_result.proto` | +fields 26, 27 (`make proto-all`) |
| `goal_result.proto` | +fields 4, 5 on `Summary` (`make proto-all`) |
| `experiment_calculator.go` | Wire `fillSequentialBayesFactors` after expected-loss; update `calculateSummary` to populate `CvrSafeToStop`/`ValueSafeToStop` |
| `experiment_calculator_test.go` | `TestCalculateSummary_SafeToStopFlags` (4 sub-cases) |
| `test/e2e/…/eventcounter_test.go` | Assert baseline BF=1, treatment CVR BF<1 (100% CVR tie), value BF>>20 (decisive fixture), `CvrSafeToStop=false`, `ValueSafeToStop=true` |

## Test plan

- [x] `go test ./pkg/experimentcalculator/experimentcalc/...` — all non-Stan unit tests pass (run locally without a Stan server)
- [x] FPR simulation passes: averaged BF FPR ≤ 6.5% (CVR), ≤ 7% (value); naive 0.95-threshold FPR is substantially higher in both
- [x] e2e `TestExperimentResult` / `TestGrpcExperimentResult` green against dev-container (requires Stan + full stack)

### CODE REVIEW

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+// models whose priors are sensitive to the distance between μ₀ and the data.
+// The always-valid FPR guarantee requires valueBFEffectSizeScale to be a
+// fixed constant — never derive it from experiment data.
+//
+// Returns 1.0 when either arm has no data, the SE is zero (no within-arm
+// variance), or any computation is non-finite.
+func valueBayesFactor(nA int64, meanA, varA float64, nB int64, meanB, varB float64) float64 {
+	if nA <= 0 || nB <= 0 {
+		return 1.0
+	}
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go:150]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/7b657942267fcdbe95f47cdd40b46ae9ab229323

File: proto/eventcounter/goal_result.proto [hunk 2]
```diff
@@ -38,4 +38,13 @@ message Summary {
   // chart (Value/User or Value/Total). best_variations (field 1) remains the
   // conversion-rate list.
   repeated Variation best_variations_value = 3;
+  // cvr_safe_to_stop is true when at least one treatment variation's
+  // cvr_sequential_bayes_factor >= DefaultSequentialBFThreshold (20). When
+  // true the sequential evidence criterion is met for CVR: the result is
+  // robust to the optional-stopping / peeking pattern.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply proto/eventcounter/goal_result.proto]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/7b657942267fcdbe95f47cdd40b46ae9ab229323

File: api-description/apidocs.swagger.yaml [hunk 3]
```diff
@@ -5702,6 +5702,18 @@ definitions:
           confidence banner can stay metric-aware when the user is viewing a value
           chart (Value/User or Value/Total). best_variations (field 1) remains the
           conversion-rate list.
+      cvrSafeToStop:
+        type: boolean
+        description: |-
+          cvr_safe_to_stop is true when at least one treatment variation's
+          cvr_sequential_bayes_factor >= DefaultSequentialBFThreshold (20). When
+          true the sequential evidence criterion is met for CVR: the result is
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/apidocs.swagger.yaml]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/7b657942267fcdbe95f47cdd40b46ae9ab229323

File: api-description/web-api.swagger.yaml [hunk 4]
```diff
@@ -10650,6 +10650,18 @@ definitions:
           confidence banner can stay metric-aware when the user is viewing a value
           chart (Value/User or Value/Total). best_variations (field 1) remains the
           conversion-rate list.
+      cvrSafeToStop:
+        type: boolean
+        description: |-
+          cvr_safe_to_stop is true when at least one treatment variation's
+          cvr_sequential_bayes_factor >= DefaultSequentialBFThreshold (20). When
+          true the sequential evidence criterion is met for CVR: the result is
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/web-api.swagger.yaml]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/7b657942267fcdbe95f47cdd40b46ae9ab229323

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+		// matching the guard in calcGoalResult).
+		if goalN > 0 && mean != 0 && variance != 0 &&
+			baseGoalN > 0 && baseMean != 0 && baseVar != 0 {
+			vr.ValueSequentialBayesFactor = valueBayesFactor(
+				goalN, mean, variance,
+				baseGoalN, baseMean, baseVar,
+			)
+		} else {
+			vr.ValueSequentialBayesFactor = 1.0
+		}
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go]: `copyDistributionSummary(nil)` (used in appendVariationResult) always returns a non-nil &DistributionSummary{}, so the nil check is always true and adds no guard. The real semantic guard is already present: mean != 0 && variance != 0 exactly mirrors calcGoalResult's skip condition. Adding the nil check would be misleading — it looks protective but does nothing.

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+	nEff := float64(nA) * float64(nB) / float64(nA+nB)
+	r := valueBFEffectSizeScale
+	nEffR2 := nEff * r * r
+
+	logBF := 0.5*math.Log(1/(1+nEffR2)) + t*t*nEffR2/(2*(1+nEffR2))
+	if !isFiniteFloat(logBF) {
+		return 1.0
+	}
+	return math.Exp(logBF)
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/changes/10a0a6631584adfe8fc9377637e94636a2a4696a

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+	}
+	// logBF = log B(sA+1, nA-sA+1) + log B(sB+1, nB-sB+1) - log B(sA+sB+1, nA+nB-sA-sB+1)
+	// Computed in log-space via lgamma to avoid overflow for large n.
+	logBF := logBetaFn(float64(sA)+1, float64(nA-sA)+1) +
+		logBetaFn(float64(sB)+1, float64(nB-sB)+1) -
+		logBetaFn(float64(sA+sB)+1, float64(nA+nB-sA-sB)+1)
+	if math.IsNaN(logBF) || math.IsInf(logBF, 0) {
+		return 1.0
+	}
+	return math.Exp(logBF)
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/changes/10a0a6631584adfe8fc9377637e94636a2a4696a

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor_test.go [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+	pB := float64(sB) / float64(nB)
+	varA := pA * (1 - pA) / float64(nA)
+	varB := pB * (1 - pB) / float64(nB)
+	se := math.Sqrt(varA + varB)
+	if se == 0 {
+		if pB > pA {
+			return 1.0
+		}
+		return 0.0
+	}
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor_test.go:611]: (se==0 → return 0.5 when pA==pB) — skip. This is a test-only demonstration helper used only to show that the naive rule inflates FPR. With θ=0.01–0.50 and 50 users/day, equal proportions are rare. The BF result is completely unaffected. Not worth touching production-adjacent test logic for an edge case in a comparison baseline.

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor_test.go [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+//
+// Baseline comparison (demonstrates the problem we are solving):
+//   The naive rule "stop when P(θ_B > θ_A | data) > 0.95" is simulated with
+//   a Normal approximation to the Beta posteriors (accurate for n >> 1).
+//   Its FPR should be substantially inflated relative to the BF rule,
+//   confirming that the BF rule is strictly better at controlling FPR.
+// ---------------------------------------------------------------------------
+
+func TestSequentialBFControlsOptionalStoppingFPR(t *testing.T) {
+	const (
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor_test.go:487]: (testing.Short() skip) — skip. Both simulations complete in ~1.4 s total (measured). go test -short is intended for tests that take minutes, and these are the correctness acceptance tests — skipping them in short mode defeats the purpose.

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor_test.go [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+	norm := distuv.Normal{Mu: 0, Sigma: 1}
+	return norm.CDF(z)
+}
+
+// ---------------------------------------------------------------------------
+// Value-metric multi-θ simulation
+// ---------------------------------------------------------------------------
+
+func TestValueBFControlsOptionalStoppingFPR(t *testing.T) {
+	const (
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor_test.go:622]: (testing.Short() skip) — skip. Both simulations complete in ~1.4 s total (measured). go test -short is intended for tests that take minutes, and these are the correctness acceptance tests — skipping them in short mode defeats the purpose.

File: api-description/apidocs.swagger.yaml [hunk 11]
```diff
@@ -7562,6 +7576,22 @@ definitions:
...
+          stopping threshold) when there is strong evidence that the CVR of
+          this variation differs from the baseline. See Summary.cvr_safe_to_stop.
+      valueSequentialBayesFactor:
+        type: number
+        format: double
+        description: |-
+          Sequential Bayes Factor for value-per-user vs baseline at the current
+          look. Always 1.0 for the baseline variation or when value data are absent.
+          Reaches 20.0 (the default stopping threshold) when there is strong
+          evidence of a value-per-user difference. See Summary.value_safe_to_stop.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/apidocs.swagger.yaml]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/a778dd9314dc1cd18b45f4deb4fe710bfc9bf555

File: api-description/apidocs.swagger.yaml [hunk 12]
```diff
@@ -7562,6 +7576,22 @@ definitions:
...
           type: number
           format: double
+      cvrSequentialBayesFactor:
+        type: number
+        format: double
+        description: |-
+          Sequential Bayes Factor for CVR vs baseline at the current look.
+          Always 1.0 for the baseline variation. Reaches 20.0 (the default
+          stopping threshold) when there is strong evidence that the CVR of
+          this variation differs from the baseline. See Summary.cvr_safe_to_stop.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/apidocs.swagger.yaml:7586]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/a778dd9314dc1cd18b45f4deb4fe710bfc9bf555

File: api-description/web-api.swagger.yaml [hunk 13]
```diff
@@ -12024,6 +12038,22 @@ definitions:
...
+          stopping threshold) when there is strong evidence that the CVR of
+          this variation differs from the baseline. See Summary.cvr_safe_to_stop.
+      valueSequentialBayesFactor:
+        type: number
+        format: double
+        description: |-
+          Sequential Bayes Factor for value-per-user vs baseline at the current
+          look. Always 1.0 for the baseline variation or when value data are absent.
+          Reaches 20.0 (the default stopping threshold) when there is strong
+          evidence of a value-per-user difference. See Summary.value_safe_to_stop.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/web-api.swagger.yaml]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/a778dd9314dc1cd18b45f4deb4fe710bfc9bf555

File: proto/eventcounter/variation_result.proto [hunk 14]
```diff
@@ -47,4 +47,14 @@ message VariationResult {
...
   repeated double cvr_samples = 25;
+  // Sequential Bayes Factor for CVR vs baseline at the current look.
+  // Always 1.0 for the baseline variation. Reaches 20.0 (the default
+  // stopping threshold) when there is strong evidence that the CVR of
+  // this variation differs from the baseline. See Summary.cvr_safe_to_stop.
+  double cvr_sequential_bayes_factor = 26;
+  // Sequential Bayes Factor for value-per-user vs baseline at the current
+  // look. Always 1.0 for the baseline variation or when value data are absent.
+  // Reaches 20.0 (the default stopping threshold) when there is strong
+  // evidence of a value-per-user difference. See Summary.value_safe_to_stop.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply proto/eventcounter/variation_result.proto]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/a778dd9314dc1cd18b45f4deb4fe710bfc9bf555

File: proto/eventcounter/goal_result.proto [hunk 15]
```diff
@@ -38,4 +38,14 @@ message Summary {
...
+  // cvr_safe_to_stop is true when at least one treatment variation's
+  // cvrSequentialBayesFactor >= 20.0 (default threshold). When true the
+  // sequential evidence criterion is met for CVR: the result is robust to
+  // the optional-stopping / peeking pattern.
+  // False does NOT mean the experiment is inconclusive — it means more data
+  // are needed before a reliable CVR verdict can be made.
+  bool cvr_safe_to_stop = 4;
+  // value_safe_to_stop is the value-per-user analogue of cvr_safe_to_stop,
+  // using valueSequentialBayesFactor >= 20.0 as the threshold.
+  bool value_safe_to_stop = 5;
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply proto/eventcounter/goal_result.proto:50]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/cefeb3f9aa8f8fd7d0047f20dbb3b76a566c2def

File: proto/eventcounter/goal_result.proto [hunk 16]
```diff
@@ -38,4 +38,14 @@ message Summary {
   // chart (Value/User or Value/Total). best_variations (field 1) remains the
   // conversion-rate list.
   repeated Variation best_variations_value = 3;
+  // cvr_safe_to_stop is true when at least one treatment variation's
+  // cvrSequentialBayesFactor >= 20.0 (default threshold). When true the
+  // sequential evidence criterion is met for CVR: the result is robust to
+  // the optional-stopping / peeking pattern.
+  // False does NOT mean the experiment is inconclusive — it means more data
+  // are needed before a reliable CVR verdict can be made.
+  bool cvr_safe_to_stop = 4;
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply proto/eventcounter/goal_result.proto:47]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/cefeb3f9aa8f8fd7d0047f20dbb3b76a566c2def

File: proto/eventcounter/variation_result.proto [hunk 17]
```diff
@@ -47,4 +47,16 @@ message VariationResult {
...
+  // stopping threshold) when there is strong evidence that the CVR of
+  // this variation differs from the baseline. See Summary.cvrSafeToStop.
+  double cvr_sequential_bayes_factor = 26;
+  // Sequential Bayes Factor for value-per-user vs baseline at the current
+  // look. Always 1.0 for the baseline variation, when value analysis is
+  // skipped (e.g. zero mean/variance), or until both arms have >= 30 users
+  // (the minimum sample size for a reliable approximation). Reaches 20.0
+  // (the default stopping threshold) when there is strong evidence of a
+  // value-per-user difference. See Summary.valueSafeToStop.
+  double value_sequential_bayes_factor = 27;
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply proto/eventcounter/variation_result.proto:62]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/cefeb3f9aa8f8fd7d0047f20dbb3b76a566c2def

File: proto/eventcounter/variation_result.proto [hunk 18]
```diff
@@ -47,4 +47,16 @@ message VariationResult {
   double conversion_rate = 23;
   double expected_loss = 24;
   repeated double cvr_samples = 25;
+  // Sequential Bayes Factor for CVR vs baseline at the current look.
+  // Always 1.0 for the baseline variation. Reaches 20.0 (the default
+  // stopping threshold) when there is strong evidence that the CVR of
+  // this variation differs from the baseline. See Summary.cvrSafeToStop.
+  double cvr_sequential_bayes_factor = 26;
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply proto/eventcounter/variation_result.proto:54]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/cefeb3f9aa8f8fd7d0047f20dbb3b76a566c2def

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+		// matching the guard in calcGoalResult).
+		if goalN > 0 && mean != 0 && variance != 0 &&
+			baseGoalN > 0 && baseMean != 0 && baseVar != 0 {
+			vr.ValueSequentialBayesFactor = valueBayesFactor(
+				goalN, mean, variance,
+				baseGoalN, baseMean, baseVar,
+			)
+		} else {
+			vr.ValueSequentialBayesFactor = 1.0
+		}
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/cefeb3f9aa8f8fd7d0047f20dbb3b76a566c2def

File: api-description/apidocs.swagger.yaml [hunk 20]
```diff
@@ -7562,6 +7576,25 @@ definitions:
...
+        type: number
+        format: double
+        description: |-
+          Sequential Bayes Factor for value-per-user vs baseline at the current
+          look. Always 1.0 for the baseline variation, when value analysis is
+          skipped (e.g. zero mean/variance for any variation in the goal), or
+          until both arms have >= 30 users (the minimum sample size for a reliable
+          approximation). Reaches 20.0 (the default stopping threshold) when there
+          is strong evidence of a value-per-user difference.
+          See Summary.value_safe_to_stop.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/apidocs.swagger.yaml]: The swagger files are auto-generated from proto. The proto comment text is the single source; changing the swagger directly would be overwritten on the next make proto-all. These comments are going in circles between camelCase and snake_case — the right answer is to pick one convention in the proto and stop.

File: api-description/apidocs.swagger.yaml [hunk 21]
```diff
@@ -7562,6 +7576,25 @@ definitions:
...
           type: number
           format: double
+      cvrSequentialBayesFactor:
+        type: number
+        format: double
+        description: |-
+          Sequential Bayes Factor for CVR vs baseline at the current look.
+          Always 1.0 for the baseline variation. Reaches 20.0 (the default
+          stopping threshold) when there is strong evidence that the CVR of
+          this variation differs from the baseline. See Summary.cvr_safe_to_stop.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/apidocs.swagger.yaml:7586]: The swagger files are auto-generated from proto. The proto comment text is the single source; changing the swagger directly would be overwritten on the next make proto-all. These comments are going in circles between camelCase and snake_case — the right answer is to pick one convention in the proto and stop.

File: api-description/apidocs.swagger.yaml [hunk 22]
```diff
@@ -5702,6 +5702,20 @@ definitions:
...
+          cvr_sequential_bayes_factor >= 20.0 (default threshold). When true the
+          sequential evidence criterion is met for CVR: the result is robust to
+          the optional-stopping / peeking pattern.
+          False does NOT mean the experiment is inconclusive — it means more data
+          are needed before a reliable CVR verdict can be made.
+      valueSafeToStop:
+        type: boolean
+        description: |-
+          value_safe_to_stop is the value-per-user analogue of cvr_safe_to_stop,
+          using value_sequential_bayes_factor >= 20.0 as the threshold.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/apidocs.swagger.yaml:5718]: The swagger files are auto-generated from proto. The proto comment text is the single source; changing the swagger directly would be overwritten on the next make proto-all. These comments are going in circles between camelCase and snake_case — the right answer is to pick one convention in the proto and stop.

File: api-description/apidocs.swagger.yaml [hunk 23]
```diff
@@ -5702,6 +5702,20 @@ definitions:
...
           chart (Value/User or Value/Total). best_variations (field 1) remains the
           conversion-rate list.
+      cvrSafeToStop:
+        type: boolean
+        description: |-
+          cvr_safe_to_stop is true when at least one treatment variation's
+          cvr_sequential_bayes_factor >= 20.0 (default threshold). When true the
+          sequential evidence criterion is met for CVR: the result is robust to
+          the optional-stopping / peeking pattern.
+          False does NOT mean the experiment is inconclusive — it means more data
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/apidocs.swagger.yaml:5712]: The swagger files are auto-generated from proto. The proto comment text is the single source; changing the swagger directly would be overwritten on the next make proto-all. These comments are going in circles between camelCase and snake_case — the right answer is to pick one convention in the proto and stop.

File: api-description/web-api.swagger.yaml [hunk 24]
```diff
@@ -12024,6 +12038,25 @@ definitions:
...
+        type: number
+        format: double
+        description: |-
+          Sequential Bayes Factor for value-per-user vs baseline at the current
+          look. Always 1.0 for the baseline variation, when value analysis is
+          skipped (e.g. zero mean/variance for any variation in the goal), or
+          until both arms have >= 30 users (the minimum sample size for a reliable
+          approximation). Reaches 20.0 (the default stopping threshold) when there
+          is strong evidence of a value-per-user difference.
+          See Summary.value_safe_to_stop.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/web-api.swagger.yaml]: The swagger files are auto-generated from proto. The proto comment text is the single source; changing the swagger directly would be overwritten on the next make proto-all. These comments are going in circles between camelCase and snake_case — the right answer is to pick one convention in the proto and stop.

File: api-description/web-api.swagger.yaml [hunk 25]
```diff
@@ -12024,6 +12038,25 @@ definitions:
...
           type: number
           format: double
+      cvrSequentialBayesFactor:
+        type: number
+        format: double
+        description: |-
+          Sequential Bayes Factor for CVR vs baseline at the current look.
+          Always 1.0 for the baseline variation. Reaches 20.0 (the default
+          stopping threshold) when there is strong evidence that the CVR of
+          this variation differs from the baseline. See Summary.cvr_safe_to_stop.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/web-api.swagger.yaml:12048]: The swagger files are auto-generated from proto. The proto comment text is the single source; changing the swagger directly would be overwritten on the next make proto-all. These comments are going in circles between camelCase and snake_case — the right answer is to pick one convention in the proto and stop.

File: api-description/web-api.swagger.yaml [hunk 26]
```diff
@@ -10650,6 +10650,20 @@ definitions:
...
+          cvr_sequential_bayes_factor >= 20.0 (default threshold). When true the
+          sequential evidence criterion is met for CVR: the result is robust to
+          the optional-stopping / peeking pattern.
+          False does NOT mean the experiment is inconclusive — it means more data
+          are needed before a reliable CVR verdict can be made.
+      valueSafeToStop:
+        type: boolean
+        description: |-
+          value_safe_to_stop is the value-per-user analogue of cvr_safe_to_stop,
+          using value_sequential_bayes_factor >= 20.0 as the threshold.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/web-api.swagger.yaml:10666]: The swagger files are auto-generated from proto. The proto comment text is the single source; changing the swagger directly would be overwritten on the next make proto-all. These comments are going in circles between camelCase and snake_case — the right answer is to pick one convention in the proto and stop.

File: api-description/web-api.swagger.yaml [hunk 27]
```diff
@@ -10650,6 +10650,20 @@ definitions:
...
           chart (Value/User or Value/Total). best_variations (field 1) remains the
           conversion-rate list.
+      cvrSafeToStop:
+        type: boolean
+        description: |-
+          cvr_safe_to_stop is true when at least one treatment variation's
+          cvr_sequential_bayes_factor >= 20.0 (default threshold). When true the
+          sequential evidence criterion is met for CVR: the result is robust to
+          the optional-stopping / peeking pattern.
+          False does NOT mean the experiment is inconclusive — it means more data
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply api-description/web-api.swagger.yaml:10660]: The swagger files are auto-generated from proto. The proto comment text is the single source; changing the swagger directly would be overwritten on the next make proto-all. These comments are going in circles between camelCase and snake_case — the right answer is to pick one convention in the proto and stop.

File: pkg/experimentcalculator/experimentcalc/experiment_calculator_test.go [hunk 28]
```diff
@@ -593,3 +593,135 @@ func TestPickBestVariations_NoneAboveThreshold(t *testing.T) {
...
+		}
+		calc.calculateSummary(context.Background(), goalResult, "baseline")
+
+		assert.False(t, goalResult.Summary.CvrSafeToStop,
+			"baseline-only result must have CvrSafeToStop=false")
+		assert.False(t, goalResult.Summary.ValueSafeToStop,
+			"baseline-only result must have ValueSafeToStop=false")
+	})
+
+	t.Run("safe_to_stop_true_when_prob_high_but_bf_is_decisive", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/experiment_calculator_test.go]: https://github.com/bucketeer-io/bucketeer/pull/2664/commits/6545b900e12fc54bbe4ee04800e1301e3099d1cd

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go [hunk 29]
```diff
@@ -0,0 +1,308 @@
...
+		if vr == nil || vr.ExperimentCount == nil {
+			continue
+		}
+		if vr.ExperimentCount.UserCount > 0 &&
+			(vr.ExperimentCount.ValueSumPerUserMean == 0 ||
+				vr.ExperimentCount.ValueSumPerUserVariance == 0) {
+			valueEnabled = false
+			break
+		}
+	}
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go:234]: https://github.com/bucketeer-io/bucketeer/pull/2664/commits/6545b900e12fc54bbe4ee04800e1301e3099d1cd

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go [hunk 30]
```diff
@@ -0,0 +1,309 @@
...
+	// variation has zero/empty value sufficient statistics (including
+	// UserCount==0). Mirror that all-or-nothing guard here: only compute
+	// value BFs when every variation has non-zero user count, mean, and
+	// variance. This prevents a decisive value BF for one arm from setting
+	// ValueSafeToStop=true in a multi-arm experiment where another arm's
+	// value posterior was skipped.
+	valueEnabled := true
+	for _, vr := range vrs {
+		if vr == nil || vr.ExperimentCount == nil {
+			continue
```
Comment (Ubisoft-potato) [inline pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go]: When `ExperimentCount == nil`, this `continue` silently skips the variation instead of disabling value BF for the entire goal. If another treatment arm has valid data, its `ValueSequentialBayesFactor` will still be computed and can trigger `ValueSafeToStop=true` — even though this arm has no data at all.

This can happen in practice: `calcGoalResult` returns early (e.g. one variation missing from `evalVariationCounts`), leaving some entries in `goalResult.VariationResults` with `ExperimentCount == nil` after `appendVariationResult`.

Fix:
```go
if vr == nil || vr.ExperimentCount == nil {
    valueEnabled = false
    break
}
```

Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor.go]: Nice catch!
Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/41e4eb3373ab4c1c2894e1f27c6e58e6c032a2d8.

File: proto/eventcounter/variation_result.proto [hunk 31]
```diff
@@ -47,4 +47,17 @@ message VariationResult {
...
+  // Sequential Bayes Factor for CVR vs baseline at the current look.
+  // Always 1.0 for the baseline variation. Reaches 20.0 (the default
+  // stopping threshold) when there is strong evidence that the CVR of
+  // this variation differs from the baseline. See Summary.cvr_safe_to_stop.
+  double cvr_sequential_bayes_factor = 26;
+  // Sequential Bayes Factor for value-per-user vs baseline at the current
+  // look. Always 1.0 for the baseline variation, when value analysis is
+  // skipped (e.g. zero mean/variance for any variation in the goal), or
+  // until both arms have >= 30 users (the minimum sample size for a reliable
+  // approximation). Reaches 20.0 (the default stopping threshold) when there
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply proto/eventcounter/variation_result.proto]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/41e4eb3373ab4c1c2894e1f27c6e58e6c032a2d8.

File: pkg/experimentcalculator/experimentcalc/sequential_bayes_factor_test.go [hunk 32]
```diff
@@ -0,0 +1,666 @@
...
+	t.Parallel()
+	assert.Equal(t, 1.0, valueBayesFactor(0, 10.0, 1.0, 10, 10.0, 1.0))
+	assert.Equal(t, 1.0, valueBayesFactor(10, 10.0, 1.0, 0, 10.0, 1.0))
+}
+
+func TestValueBayesFactor_EqualArms_NullFavoured(t *testing.T) {
+	t.Parallel()
+	// Both arms identical — data should favour H₀ → BF < 1.
+	bf := valueBayesFactor(100, 10.0, 2.0, 100, 10.0, 2.0)
+	assert.True(t, bf < 1.0, "equal arms should give BF < 1 (H₀ favoured), got %f", bf)
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/experimentcalculator/experimentcalc/sequential_bayes_factor_test.go]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2664/commits/41e4eb3373ab4c1c2894e1f27c6e58e6c032a2d8.

**Review by Ubisoft-potato — APPROVED** [review]
Nice! 👍 
Thanks!
